src/workerd/api/tests/js-rpc-test.js

```diff
@@ -832,6 +860,118 @@ export let serializeRpcPromiseOrProprety = {
 }
 
+export let streams = {
+  async test(controller, env, ctx) {
+    // Send WritableStream.
+    {
+      let { readable, writable } = new IdentityTransformStream();
+      let promise = env.MyService.writeToStream(writable);
+      let text = await new Response(readable).text();
+      assert.strictEqual(text, "foo, bar, baz!");
+      await promise;
+    }
+
+    // TODO(someday): Test JS-backed WritableStream, when it actually works.
+
+    // TODO(someday): Is there any way to construct an encoded WritableStream? Only system
+    //   streams can be encoded, but there's no API that returns an encoded WritableStream I think.
+
+    // Send ReadableStream.
+    {
+      let { readable, writable } = new IdentityTransformStream();
+      let promise = env.MyService.readFromStream(readable);
+
+      let writer = writable.getWriter();
+      let enc = new TextEncoder();
+      await writer.write(enc.encode("foo, "));
+      await writer.write(enc.encode("bar, "));
+      await writer.write(enc.encode("baz!"));
+      await writer.close();
+
+      assert.strictEqual(await promise, "foo, bar, baz!");
+    }
+
+    // Send a JS-backed ReadableStream.
+    {
+      let controller;
+      let readable = new ReadableStream({
+        start(c) { controller = c; }
+      });
+      let promise = env.MyService.readFromStream(readable);
+
+      let enc = new TextEncoder();
+      controller.enqueue(enc.encode("foo, "));
+      controller.enqueue(enc.encode("bar, "));
+      controller.enqueue(enc.encode("baz!"));
+      controller.close();
+
+      assert.strictEqual(await promise, "foo, bar, baz!");
+    }
+
+    // Send a ReadableStream that is locked.
+    {
+      let { readable, writable } = new IdentityTransformStream();
+
+      let reader = readable.getReader();
+
+      assert.rejects(env.MyService.readFromStream(readable), {
+        name: "TypeError",
+        message: "The ReadableStream has been locked to a reader."
+      });
+    }
+
+    // Receive ReadableStream.
+    {
+      let readable = await env.MyService.returnReadableStream();
+      let text = await new Response(readable).text();
+      assert.strictEqual(text, "foo, bar, baz!");
+    }
+
+    // Send ReadableStream, but fail to fully write it.
+    {
+      let { readable, writable } = new IdentityTransformStream();
+      let promise = env.MyService.readFromStream(readable);
+
+      let writer = writable.getWriter();
+      let enc = new TextEncoder();
+      await writer.write(enc.encode("foo, "));
+      await writer.write(enc.encode("bar, "));
+      await writer.write(enc.encode("baz!"));
+      await writer.abort("foo");
+
+      await assert.rejects(promise, {
+        name: "Error",
+        // TODO(someday): Propagate the actual error.
+        message: "ReadableStream received over RPC disconnected prematurely."
+      });
+    }
+
+    // Send fixed-length ReadableStream.
+    {
+      let { readable, writable } = new FixedLengthStream("foo, bar, baz!".length);
```

Comment:
non-blocking: a test that passes the `readable` or `writable` as a nested property of some other object would be good also here. This would demonstrate that the value serializer / deserializer delegate is properly handling nested custom api types. If it didn't, that would more signal an issue in v8 than our code, but certainly a good thing to verify.

Also, it strikes me that someone could create a `TransformStream` on one side of the RPC channel, and pass both the `readable` and `writable` to the other side such that transformations are occurring across the RPC boundary but the main stream operations are happening locally... e.g.

```
const { readable, writable } = env.MyService.GetTransformer();
```

That's actually pretty nice.

Reply:
There are a number of existing test cases already for nested properties (using RpcTarget, functions, etc.). However, a test that passes multiple ReadableStreams at the same time is a good idea, to make sure the right indexes are passed to StreamSink. And indeed this caught a bug: 26f332f0